[WTF] Introduce StatisticsManager for recording per-key numeric statistics with histogram support

hyjorc1 opened the pull request:
#### 222d2e04c1d686c2294db4a68733d64da28825b3
<pre>
[WTF] Introduce StatisticsManager for recording per-key numeric statistics with histogram support
<a href="https://bugs.webkit.org/show_bug.cgi?id=291647">https://bugs.webkit.org/show_bug.cgi?id=291647</a>
<a href="https://rdar.apple.com/149416894">rdar://149416894</a>

Reviewed by Keith Miller.

This patch introduces `WTF::StatisticsManager`, a thread-safe singleton for
recording statistical metrics (e.g. count, min, max, mean, variance, standard
deviation) for arbitrary string identifiers.

Each identifier (e.g. &quot;unrolled-loop-loopBodySize&quot;) accumulates the
full data history (Vector&lt;double&gt;) in time. And count, sum, sum-of-squares,
min, max, and histogram with 10 bins, computed on-demand during dumpStatistics().

This supports runtime use cases such as performance profiling, optimization
validation, and internal monitoring.

Example usage:

    StatisticsManager::singleton().addDataPoint(&quot;fully-unrolled-loop-loopBodySize&quot;, loopSize);
    ...
    StatisticsManager::singleton().addDataPoint(&quot;partially-unrolled-loop-loopBodySize&quot;, loopSize);

    ...
    StatisticsManager::singleton().dumpStatistics();
    StatisticsManager::singleton().clear(); // optional

Output format:

Statistics for fully-unrolled-loop-loopBodySize:
    Count    : 34
    Min      : 22.000000
    Max      : 189.000000
    Mean     : 49.794118
    Variance : 1487.045848
    Std Dev  : 38.562233
    Histogram:
    [22.00 - 38.70): 21
    [38.70 - 55.40): 6
    [55.40 - 72.10): 2
    [72.10 - 88.80): 1
    [88.80 - 105.50): 0
    [105.50 - 122.20): 1
    [122.20 - 138.90): 1
    [138.90 - 155.60): 1
    [155.60 - 172.30): 0
    [172.30 - 189.00): 1

Statistics for partially-unrolled-loop-loopBodySize:
    ...

Future extensions may include percentile estimation, histogram range controls,
or scoped/statistical macros for easier instrumentation.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StatisticsManager.cpp: Added.
(WTF::StatisticsManager::singleton):
(WTF::StatisticsManager::StatBucket::add):
(WTF::StatisticsManager::StatBucket::mean const):
(WTF::StatisticsManager::StatBucket::variance const):
(WTF::StatisticsManager::StatBucket::stdDev const):
(WTF::StatisticsManager::StatBucket::prepareHistogram):
(WTF::StatisticsManager::StatBucket::addToHistogram):
(WTF::StatisticsManager::StatBucket::dumpHistogram):
(WTF::StatisticsManager::addDataPoint):
(WTF::StatisticsManager::dumpStatistics):
(WTF::StatisticsManager::clear):
* Source/WTF/wtf/StatisticsManager.h: Added.

Canonical link: <a href="https://commits.webkit.org/293856@main">https://commits.webkit.org/293856@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/584f191896f3fb50aa495feff045ead207f7a9eb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100121 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19769 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10066 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105251 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50702 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28242 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76216 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33283 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103128 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90425 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56577 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8423 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50072 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92779 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107610 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98728 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27234 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86632 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84706 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21519 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29383 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21089 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27171 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32403 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122354 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26982 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34151 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30298 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28541 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/WTF/wtf/StatisticsManager.h:
```diff
@@ -0,0 +1,80 @@
+        double max { -std::numeric_limits<double>::infinity() };
+
+        Vector<double> allValues;
+        Vector<size_t, numBuckets> histogram;
```
It looks like "bucket" terminology is being overloaded to mean both the thing that holds the stats for a key and also the thing that makes up a histogram. I'd rename the former to disambiguate.

hyjorc1 replied:
Fixed in the post aggregation.

dhecht commented on Source/WTF/wtf/StatisticsManager.h:
```diff
@@ -0,0 +1,80 @@
+        void add(double value);
+        double mean() const;
+        double variance() const;
+        double stdDev() const;
```
maybe pass `AbstractLocker&` to mean, variance, stdDev etc to document that the caller should be holding the lock.

hyjorc1 replied:
Fixed in the post aggregation.

kmiller68 commented on Source/WTF/wtf/StatisticsManager.h:
```diff
@@ -0,0 +1,52 @@
+    ~StatisticsManager() = default;
+
+    mutable Lock m_lock;
+    UncheckedKeyHashMap<String, Vector<double>> m_data WTF_GUARDED_BY_LOCK(m_lock);
```
This seems fine for now but why not a `UncheckedKeyHashMap<String, std::unique_ptr<HashMap<double, size_t>>>`? Also, this seems weird because you're passing a String (not thread safe yet) in but it has a lock?

Maybe this should be an `ASCIILiteral` or a `std::string` for now.

hyjorc1 replied:
Thanks. Updated with `ASCIILiteral`. But for `std::unique_ptr<HashMap<double, size_t>>`, even it can reduced memory usage if the values repeat often, it's not ideal for future extensions because: 1. double as key is fragile. 2. hard to do range-based analysis (binning, percentiles, etc.). 3. Not flexible for future needs.

kmiller68 reviewed (approved):
r=me

hyjorc1 committed:
[WTF] Introduce StatisticsManager for recording per-key numeric statistics with histogram support
https://bugs.webkit.org/show_bug.cgi?id=291647
rdar://149416894

Reviewed by Keith Miller.

This patch introduces `WTF::StatisticsManager`, a thread-safe singleton for
recording statistical metrics (e.g. count, min, max, mean, variance, standard
deviation) for arbitrary string identifiers.

Each identifier (e.g. "unrolled-loop-loopBodySize") accumulates the
full data history (Vector<double>) in time. And count, sum, sum-of-squares,
min, max, and histogram with 10 bins, computed on-demand during dumpStatistics().

This supports runtime use cases such as performance profiling, optimization
validation, and internal monitoring.

Example usage:

    StatisticsManager::singleton().addDataPoint("fully-unrolled-loop-loopBodySize", loopSize);
    ...
    StatisticsManager::singleton().addDataPoint("partially-unrolled-loop-loopBodySize", loopSize);

    ...
    StatisticsManager::singleton().dumpStatistics();
    StatisticsManager::singleton().clear(); // optional

Output format:

Statistics for fully-unrolled-loop-loopBodySize:
    Count    : 34
    Min      : 22.000000
    Max      : 189.000000
    Mean     : 49.794118
    Variance : 1487.045848
    Std Dev  : 38.562233
    Histogram:
    [22.00 - 38.70): 21
    [38.70 - 55.40): 6
    [55.40 - 72.10): 2
    [72.10 - 88.80): 1
    [88.80 - 105.50): 0
    [105.50 - 122.20): 1
    [122.20 - 138.90): 1
    [138.90 - 155.60): 1
    [155.60 - 172.30): 0
    [172.30 - 189.00): 1

Statistics for partially-unrolled-loop-loopBodySize:
    ...

Future extensions may include percentile estimation, histogram range controls,
or scoped/statistical macros for easier instrumentation.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StatisticsManager.cpp: Added.
(WTF::StatisticsManager::singleton):
(WTF::StatisticsManager::StatBucket::add):
(WTF::StatisticsManager::StatBucket::mean const):
(WTF::StatisticsManager::StatBucket::variance const):
(WTF::StatisticsManager::StatBucket::stdDev const):
(WTF::StatisticsManager::StatBucket::prepareHistogram):
(WTF::StatisticsManager::StatBucket::addToHistogram):
(WTF::StatisticsManager::StatBucket::dumpHistogram):
(WTF::StatisticsManager::addDataPoint):
(WTF::StatisticsManager::dumpStatistics):
(WTF::StatisticsManager::clear):
* Source/WTF/wtf/StatisticsManager.h: Added.

Canonical link: https://commits.webkit.org/293856@main
